internal/status/cache.go

```diff
@@ -16,27 +16,41 @@
 package status
 
 import (
-	"time"
-
 	contour_api_v1 "github.com/projectcontour/contour/apis/projectcontour/v1"
 	"github.com/projectcontour/contour/internal/k8s"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
 )
 
-type ProxyStatus string
+// ConditionType is used to ensure we only use a limited set of possible values
+// for DetailedCondition types. It's cast back to a string before sending off to
+// HTTPProxy structs, as those use upstream types which we can't alias easily.
+type ConditionType string
 
-const (
-	ProxyStatusValid    ProxyStatus = "valid"
-	ProxyStatusInvalid  ProxyStatus = "invalid"
-	ProxyStatusOrphaned ProxyStatus = "orphaned"
-)
+// ValidCondition is the ConditionType for Valid.
+const ValidCondition ConditionType = "Valid"
+
+// AdmittedConditionValid is the valid ConditionType for Admitted.
+const AdmittedConditionValid ConditionType = "AdmittedValid"
+
+// AdmittedConditionInValid is the invalid ConditionType for Admitted.
+const AdmittedConditionInvalid ConditionType = "AdmittedInvalid"
+
+// NotImplemented is the ConditionType for NotImplemented.
+const NotImplemented ConditionType = "NotImplemented"
+
+// Invalid is the ConditionType for Invalid.
+const Invalid ConditionType = "Invalid"
 
 // NewCache creates a new Cache for holding status updates.
-func NewCache() Cache {
+func NewCache(gateway types.NamespacedName) Cache {
 	return Cache{
-		proxyUpdates: make(map[types.NamespacedName]*ProxyUpdate),
-		entries:      make(map[string]map[types.NamespacedName]CacheEntry),
+		proxyUpdates:     make(map[types.NamespacedName]*ProxyUpdate),
+		gatewayRef:       gateway,
+		httpRouteUpdates: make(map[types.NamespacedName]*HTTPRouteUpdate),
```

Comment:
is it possible to use the generic `entries` structure to store the HTTPRoute data, like with `ExtensionService`? If I remember correctly, I think there was a desire to move to using that. See also https://github.com/projectcontour/contour/blob/main/internal/status/extensionstatus.go#L108 and https://github.com/projectcontour/contour/blob/main/internal/dag/extension_processor.go#L44

Reply:
The problem I ran into was that the `CacheEntry` interface has a `ConditionFor` method returning a `DetailedCondition` type which is specific to HTTPProxy and GatewayAPI uses a `metav1.Condition` type.

Reply:
Ah, yeah that makes sense. I'll take another look with that in mind.

Reply:
We could potentially refactor this into something similar to how it works with ExtensionService since we'll also need to set status on TLSRotues & Gateways.

Reply:
Yeah, I'm not thrilled that we now have 3 slightly different ways of doing status updates for 3 different types, so let's definitely look for opportunities to consolidate. I'd need to look in some more detail at all of them to have suggestions though.

Reply:
Probably worth filing a follow-up issue to look at cleaning some of this up.